src/t_cose_mac_verify.c

```diff
@@ -0,0 +1,335 @@
+    struct q_useful_buf_c     tag;
+    int32_t                   hash_alg_id;
+
+    hash_alg_id = t_cose_hmac_to_hash_alg_id(cose_alg_id);
+    if (hash_alg_id == INT32_MAX) {
+        return_value = T_COSE_ERR_UNSUPPORTED_SIGNING_ALG;
+        goto Done;
+    }
+
+    return_value = t_cose_crypto_hash_start(&hash_ctx, hash_alg_id);
+    if (return_value != T_COSE_SUCCESS) {
+        goto Done;
+    }
+
+    /* Hash the Header */
+    t_cose_crypto_hash_update(&hash_ctx, q_useful_buf_head(header, header.len));
+
+    /* Hash the payload */
+    t_cose_crypto_hash_update(&hash_ctx, payload);
+
+    return_value = t_cose_crypto_hash_finish(&hash_ctx, tag_buffer, &tag);
+    if (return_value != T_COSE_SUCCESS) {
+        goto Done;
+    }
+
+    if (q_useful_buf_compare(tag_to_verify, tag)) {
+        return_value = T_COSE_ERR_SIG_VERIFY;
+    } else {
+        return_value = T_COSE_SUCCESS;
+    }
+
+Done:
+    return return_value;
+}
+#endif /* T_COSE_DISABLE_SHORT_CIRCUIT_SIGN */
+
+/**
+ * \file t_cose_mac_verify.c
+ *
+ * \brief This verifies t_cose Mac authentication structure without a recipient
+ *        structure.
+ *        Only HMAC is supported so far.
+ */
+
+/*
+ * Public function. See t_cose_mac.h
+ */
+enum t_cose_err_t t_cose_mac_verify(struct t_cose_mac_verify_ctx *context,
+                                     struct q_useful_buf_c        cose_mac,
+                                     struct q_useful_buf_c       *payload)
+{
+    QCBORDecodeContext            decode_context;
+    struct q_useful_buf_c         protected_parameters;
+    struct t_cose_parameters      parameters;
+    struct t_cose_label_list      critical_parameter_labels;
+    struct t_cose_label_list      unknown_parameter_labels;
+    QCBORError                    qcbor_error;
+
+    enum t_cose_err_t             return_value;
+    struct q_useful_buf_c         tag = NULL_Q_USEFUL_BUF_C;
+    struct q_useful_buf_c         tbm_first_part;
+    /* Buffer for the ToBeMaced */
+    Q_USEFUL_BUF_MAKE_STACK_UB(   tbm_first_part_buf,
+                                  T_COSE_SIZE_OF_TBM);
+    struct t_cose_crypto_hmac     hmac_ctx;
+
+    *payload = NULL_Q_USEFUL_BUF_C;
+
+    clear_label_list(&unknown_parameter_labels);
+    clear_label_list(&critical_parameter_labels);
+    clear_cose_parameters(&parameters);
+
+    QCBORDecode_Init(&decode_context, cose_mac, QCBOR_DECODE_MODE_NORMAL);
+
+    /* --- The array of 4 and tags --- */
+    QCBORDecode_EnterArray(&decode_context, NULL);
+    return_value = qcbor_decode_error_to_t_cose_error(QCBORDecode_GetError(&decode_context));
+    if(return_value != T_COSE_SUCCESS) {
+        goto Done;
+    }
+    return_value = process_tags(context, &decode_context);
+    if(return_value != T_COSE_SUCCESS) {
+        goto Done;
+    }
+
+    /* --- The protected parameters --- */
+    QCBORDecode_EnterBstrWrapped(&decode_context, QCBOR_TAG_REQUIREMENT_NOT_A_TAG, &protected_parameters);
```

Comment:
There's a whole new API for decoding header parameters in t_cose_parameters.h. This should probably switch to that to be able to handle custom parameters and so that COSE_Recipients can be part of MAC at some point.

Reply:
I can see that the sign1 verify part of the code is adjested to the now API bu not fully I mean the context structure is not updated and some other things. I got mac verify part to the same level and I am leaving it for now since I expect things might sill change and I could really use a working example.

Reply:
I saw a comment from Adam in email, but can't find it here in GitHub. 

I want to respond and say I think it is OK to merge COSE_Mac even if he parameter stuff is the old parameter stuff. If you fix the merge conflicts and get the tests running we can merge Mac.